### ISSUE TITLE
TypeVar with no constraints treated as Any in function parameters

### ISSUE THREAD

**Issue by edkek** [python/mypy#7433]
I am trying to make a function that accepts multiple parameters with the same generic type. I assumed mypy would be able to do this (since it seems PyCharm is able to detect this kind of thing), however mypy does not report this correctly it seems.

If I make a TypeVar and name it `T` by doing `T = TypeVar('T')`, and then use `T` as a type for multiple parameters in a function, it should enforce that those parameters have the same type `T`. This seems to work if the TypeVar has constraints (such as if I do `T = TypeVar('T', int, float, str)`), this doesn't work if I have no constraints, it seems to treat it as `Any`. However, it's not really `Any`, since we can infer the type `T` from the first parameter of type `T`.

```python
from typing import TypeVar

T = TypeVar('T')

def bug(a: T, b: T):
    assert type(b) == type(a)


bug(5, 5)  # No error
bug(5.0, 5)  # No error, should be ok since int is a type of float
bug("5", 5)  # No error, but should error since T should be inferred as str
```

* Are you reporting a bug, or opening a feature request?
I believe this is a bug, if I understand how generic parameters in Python work

* What are the versions of mypy and Python you are using?
`Python 3.6.8 :: Anaconda, Inc.` and `mypy 0.701`

* What are the mypy flags you are using? (For example --strict-optional)
Both `--strict` and `--strict-optional` produce the same result

**Comment by msullivan** [comment]
While I admittedly freaked out a little when I saw this issue at first, upon reflection I believe that it is not a bug, though it is perhaps surprising:

What is happening here is that `T` is being inferred as `object`, since that is the common supertype of the argument types.

**Comment by ilevkivskyi** [comment]
This is literally the first example in the PEP 483 about type variables: https://www.python.org/dev/peps/pep-0483/#type-variables. If this will appear again maybe it would help to copy this to RtD.

Why do you need such a function? Sometimes setting an upper bound helps (you can also set it to a protocol like `T = TypeVar('T', bound=Comparable)`), so that `T` will not go too high.

There is an issue about giving a warning in type variable inferred as `object` (only for builtin containers however), and there is even a (stalled) PR.

**Comment by edkek** [comment]
Ah ok this makes a lot more sense now. I was under the impression that `TypeVar` inferred the most specific type, not the most general (in this case `object`). I'll give PEP 483 another look over. Thank you for the clarification and sorry for the false flag issue
